omniqueue/src/backends/redis/mod.rs

```diff
@@ -137,6 +190,177 @@ where
+    R::Error: 'static + std::error::Error + Send + Sync,
+{
+    if delayed_queue_key.is_empty() {
+        tracing::warn!("no delayed_queue_key specified - delayed task scheduler disabled");
+        return None;
+    }
+
+    Some(tokio::spawn({
+        let pool = redis.clone();
+        let mqn = queue_key.to_string();
+        let dqn = delayed_queue_key.to_string();
+        let delayed_lock = format!("{delayed_queue_key}__lock");
+        let payload_key = payload_key.to_string();
+        tracing::debug!(
+            "spawning delayed task scheduler: delayed_queue_key=`{delayed_queue_key}`, \
+            delayed_lock=`{delayed_lock}`"
+        );
+
+        async move {
+            loop {
+                if let Err(err) = background_task_delayed(
+                    pool.clone(),
+                    mqn.clone(),
+                    dqn.clone(),
+                    &delayed_lock,
+                    &payload_key,
+                )
+                .await
+                {
+                    tracing::error!("{}", err);
+                    tokio::time::sleep(Duration::from_millis(500)).await;
+                    continue;
+                };
+            }
+        }
+    }))
+}
+
+/// Special ID for XADD command's which generates a stream ID automatically
+const GENERATE_STREAM_ID: &str = "*";
+/// Special ID for XREADGROUP commands which reads any new messages
+const LISTEN_STREAM_ID: &str = ">";
+
+/// Moves "due" messages from a sorted set, where delayed messages are shelved, back onto the main queue.
+async fn background_task_delayed<R>(
+    pool: bb8::Pool<R>,
+    main_queue_name: String,
+    delayed_queue_name: String,
+    delayed_lock: &str,
+    payload_key: &str,
+) -> Result<(), QueueError>
+where
+    R: RedisConnection,
+    R::Connection: redis::aio::ConnectionLike + Send + Sync,
+    R::Error: 'static + std::error::Error + Send + Sync,
+{
+    let batch_size: isize = 50;
+
+    let mut conn = pool.get().await.map_err(QueueError::generic)?;
+
+    // There is a lock on the delayed queue processing to avoid race conditions. So first try to
+    // acquire the lock should it not already exist. The lock expires after five seconds in case a
+    // worker crashes while holding the lock.
+    let mut cmd = redis::cmd("SET");
+    cmd.arg(delayed_lock)
+        .arg(true)
+        .arg("NX")
+        .arg("PX")
+        .arg(5000);
+    // WIll be Some("OK") when set or None when not set
+    let resp: Option<String> = cmd
+        .query_async(&mut *conn)
+        .await
+        .map_err(QueueError::generic)?;
+
+    if resp.as_deref() == Some("OK") {
+        // First look for delayed keys whose time is up and add them to the main queue
+
+        // FIXME(onelson): original impl depends on chrono so it can convert to UTC. Should probably do this too.
+        let timestamp: i64 = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
```

Comment:
Seems like this is good as-is. I verified that `SystemTime` gives UTC time, regardless of the TZ selection for the process.

Put another way, where running `env TZ=America/New_York date` shows the tz shift away from my system default of `America/Los_Angeles`, the output from `SystemTime` is stable.